### PR TITLE
Add support for a RestClient path prefix

### PR DESCRIPTION
**Pull request by pickypg**
This enables simple support for proxies (beyond proxy host and proxy port, which is done via the RequestConfig)) to provide a base path in front of all requests performed by the `RestClient`.

It also adds some Javadocs.

### CODE REVIEW

File: client/rest/src/main/java/org/elasticsearch/client/RestClient.java [hunk 1]
```diff
@@ -89,17 +89,20 @@
     //we don't rely on default headers supported by HttpAsyncClient as those cannot be replaced
     private final Header[] defaultHeaders;
     private final long maxRetryTimeoutMillis;
+    // basePath is stored in the form "/base/path" so there's never a trailing slash
+    private final String basePath;
```
Comment (javanna) [inline client/rest/src/main/java/org/elasticsearch/client/RestClient.java]: can we call it pathPrefix ?

File: client/rest/src/main/java/org/elasticsearch/client/RestClient.java [hunk 2]
```diff
@@ -89,17 +89,20 @@
     //we don't rely on default headers supported by HttpAsyncClient as those cannot be replaced
     private final Header[] defaultHeaders;
     private final long maxRetryTimeoutMillis;
+    // basePath is stored in the form "/base/path" so there's never a trailing slash
```
Comment (javanna) [inline client/rest/src/main/java/org/elasticsearch/client/RestClient.java]: do we validate this format?

File: client/rest/src/main/java/org/elasticsearch/client/RestClient.java [hunk 3]
```diff
@@ -501,10 +504,20 @@ private static HttpRequestBase addRequestBody(HttpRequestBase httpRequest, HttpE
         return httpRequest;
     }
 
-    private static URI buildUri(String path, Map<String, String> params) {
+    private static URI buildUri(String basePath, String path, Map<String, String> params) {
         Objects.requireNonNull(params, "params must not be null");
         try {
-            URIBuilder uriBuilder = new URIBuilder(path);
+            String fullPath = path;
```
Comment (javanna) [inline client/rest/src/main/java/org/elasticsearch/client/RestClient.java]: can we assign path fo fullPath only when needed?

File: client/rest/src/main/java/org/elasticsearch/client/RestClient.java [hunk 4]
```diff
@@ -501,10 +504,20 @@ private static HttpRequestBase addRequestBody(HttpRequestBase httpRequest, HttpE
...
+
+            if (basePath != null) {
+                if (path.startsWith("/") == false) {
```
Comment (javanna) [inline client/rest/src/main/java/org/elasticsearch/client/RestClient.java]: can we do the opposite? if path.startsWith then else ?

File: client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java [hunk 5]
```diff
@@ -119,14 +133,49 @@ public RestClientBuilder setRequestConfigCallback(RequestConfigCallback requestC
...
+     * <p>
+     * In essence, every request's {@code endpoint} is prefixed by this {@code basePath}. The base path is useful for when Elasticsearch is
+     * behind a proxy that provides a base path; it is not intended for other purposes.
+     *
+     * @throws NullPointerException if {@code basePath} is {@code null}.
+     * @throws IllegalArgumentException if {@code basePath} is empty, only '/', or ends with more than one '/'.
+     */
+    public RestClientBuilder setBasePath(String basePath) {
+        Objects.requireNonNull(basePath, "basePath must not be null");
+        String cleanBasePath = basePath.trim();
```
Comment (javanna) [inline client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java]: why do we trim it?

File: client/rest/src/main/java/org/elasticsearch/client/RestClient.java (hunk 2 above)
Comment (javanna) [inline client/rest/src/main/java/org/elasticsearch/client/RestClient.java]: I see that we do, sorry for the noise. maybe this comment is not needed though?

File: client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java [hunk 6]
```diff
@@ -119,14 +133,49 @@ public RestClientBuilder setRequestConfigCallback(RequestConfigCallback requestC
...
+        if (cleanBasePath.endsWith("/")) {
+            cleanBasePath = cleanBasePath.substring(0, cleanBasePath.length() - 1);
+
+            if (cleanBasePath.endsWith("/")) {
+                throw new IllegalArgumentException("basePath is malformed. too many trailing slashes: [" + basePath + "]");
+            }
+        }
+
+        if (cleanBasePath.isEmpty() || "/".equals(cleanBasePath)) {
+            // just don't set it
```
Comment (javanna) [inline client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java]: redundant comment?

File: client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java [hunk 7]
```diff
@@ -119,14 +133,49 @@ public RestClientBuilder setRequestConfigCallback(RequestConfigCallback requestC
...
+     * @throws NullPointerException if {@code basePath} is {@code null}.
+     * @throws IllegalArgumentException if {@code basePath} is empty, only '/', or ends with more than one '/'.
+     */
+    public RestClientBuilder setBasePath(String basePath) {
+        Objects.requireNonNull(basePath, "basePath must not be null");
+        String cleanBasePath = basePath.trim();
+
+        if (cleanBasePath.startsWith("/") == false) {
+            cleanBasePath = "/" + cleanBasePath;
+        }
```
Comment (javanna) [inline client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java]: question: do we need to add a leading '/' if not there? why?

File: client/rest/src/test/java/org/elasticsearch/client/RestClientBuilderTests.java [hunk 8]
```diff
@@ -103,41 +104,82 @@ public void testBuild() throws IOException {
...
         }
-        if (getRandom().nextBoolean()) {
-            builder.setMaxRetryTimeoutMillis(RandomInts.randomIntBetween(getRandom(), 1, Integer.MAX_VALUE));
+        if (randomBoolean()) {
+            builder.setMaxRetryTimeoutMillis(randomIntBetween(1, Integer.MAX_VALUE));
+        }
+        if (randomBoolean()) {
+            String basePath = randomAsciiOfLengthBetween(2, 5);
+
+            while (basePath.length() < 20 && randomBoolean()) {
```
Comment (javanna) [inline client/rest/src/test/java/org/elasticsearch/client/RestClientBuilderTests.java]: isn't basePath.length() < 20 always true?

File: client/rest/src/test/java/org/elasticsearch/client/RestClientBuilderTests.java [hunk 9]
```diff
@@ -103,41 +104,82 @@ public void testBuild() throws IOException {
...
+            builder.setBasePath(basePath);
         }
         try (RestClient restClient = builder.build()) {
             assertNotNull(restClient);
         }
     }
+
+    public void testSetBasePathNull() {
+        try {
+            RestClient.builder(new HttpHost("localhost", 9200)).setBasePath(null);
```
Comment (javanna) [inline client/rest/src/test/java/org/elasticsearch/client/RestClientBuilderTests.java]: missing fail :) use expectThrows instead

File: client/rest/src/test/java/org/elasticsearch/client/RestClientBuilderTests.java [hunk 10]
```diff
@@ -103,41 +104,82 @@ public void testBuild() throws IOException {
...
+        assertSetBasePathThrows("  /  ");
+        assertSetBasePathThrows(" / ");
+        assertSetBasePathThrows("/");
+        assertSetBasePathThrows(" ");
+        assertSetBasePathThrows("");
+    }
+
+    public void testSetBasePathMalformed() {
+        assertSetBasePathThrows("//");
+        assertSetBasePathThrows("base/path//");
```
Comment (javanna) [inline client/rest/src/test/java/org/elasticsearch/client/RestClientBuilderTests.java]: use expectThrows instead and drop the added method as each assert will be a single line without it too?

File: client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java [hunk 11]
```diff
@@ -205,6 +206,39 @@ public void testGetWithBody() throws IOException {
...
+        try (final RestClient client =
+                RestClient.builder(new HttpHost(httpServer.getAddress().getHostString(), httpServer.getAddress().getPort()))
+                    .setBasePath(basePath).build()) {
+
+            context = httpServer.createContext("/" + basePath + statusCode, new ResponseHandler(statusCode));
+
+            for (String method : getHttpMethods()) {
+                Response esResponse;
+                try {
+                    esResponse = client.performRequest(method, "/" + statusCode);
```
Comment (javanna) [inline client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java]: I am afraid that this could end up working accidentally thanks to the contexts that get registered in the before class method. Maybe this test should be better isolated and have only its own context defined?

**Comment by javanna** [comment]
left a few comments

File: client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java (hunk 6 above)
Comment (pickypg) [inline client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java]: I left that simple comment there incase a user hit the exception and searched our code. I'll remove it though because I added the Javadoc afterward.

File: client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java (hunk 7 above)
Comment (pickypg) [inline client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java]: I actually just assumed that we did to make a well-formed HTTP request. I'll test and report back.

File: client/rest/src/test/java/org/elasticsearch/client/RestClientBuilderTests.java (hunk 8 above)
Comment (pickypg) [inline client/rest/src/test/java/org/elasticsearch/client/RestClientBuilderTests.java]: Not if you keep hitting `randomBoolean() == true` "luckily". It's `basePath +=`, not `basePath = ` to keep adding random subpaths without getting too crazy.

File: client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java (hunk 7 above)
Comment (javanna) [inline client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java]: I think I recall cases that work perfectly without the leading "/". e.g. https://github.com/elastic/elasticsearch/issues/19314

File: client/rest/src/test/java/org/elasticsearch/client/RestClientBuilderTests.java (hunk 10 above)
Comment (pickypg) [inline client/rest/src/test/java/org/elasticsearch/client/RestClientBuilderTests.java]: Can't use expectThrows since it's all Java 7 compatible. :(

File: client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java (hunk 5 above)
Comment (pickypg) [inline client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java]: I was trying to help the user, but it's honestly not enough. If they put spaces in the middle, then it's just as broken.

File: client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java (hunk 11 above)
Comment (pickypg) [inline client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java]: Good point.

> Maybe this test should be better isolated and have only its own context defined?

I'll add a unique suffix so that there's no way the context exists without being given the status code _and_ something else.

Comment (javanna) [inline client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java]: cool maybe also check that the request fails without the custom context?

File: client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java (hunk 7 above)
Comment (pickypg) [inline client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java]: I'm going to drop this from here and move it into the core `RestClient`. It seems that the underlying `HttpClient` does not prefix with a leading slash, which is the underlying issue for #19314, but the RFC states that the path should be an _absolute_ path (section 5.1.2).

Comment (javanna) [inline client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java]: if we make such change, can we do it in a separate PR please? ;)

**Comment by pickypg** [comment]
@javanna updated based on comments. Thanks.

File: client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java [hunk 12]
```diff
@@ -119,14 +133,52 @@ public RestClientBuilder setRequestConfigCallback(RequestConfigCallback requestC
...
+        if (cleanPathPrefix.endsWith("/")) {
+            cleanPathPrefix = cleanPathPrefix.substring(0, cleanPathPrefix.length() - 1);
+
+            if (cleanPathPrefix.endsWith("/")) {
+                throw new IllegalArgumentException("pathPrefix is malformed. too many trailing slashes: [" + pathPrefix + "]");
+            }
+        }
+
+        if (cleanPathPrefix.isEmpty() || "/".equals(cleanPathPrefix)) {
+            // just don't set it
```
Comment (pickypg) [inline client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java]: Need to remove this line; forgot.

File: client/rest/src/main/java/org/elasticsearch/client/RestClient.java [hunk 13]
```diff
@@ -501,10 +503,22 @@ private static HttpRequestBase addRequestBody(HttpRequestBase httpRequest, HttpE
         return httpRequest;
     }
 
-    private static URI buildUri(String path, Map<String, String> params) {
+    private static URI buildUri(String pathPrefix, String path, Map<String, String> params) {
         Objects.requireNonNull(params, "params must not be null");
         try {
-            URIBuilder uriBuilder = new URIBuilder(path);
+            String fullPath;
+
```
Comment (javanna) [inline client/rest/src/main/java/org/elasticsearch/client/RestClient.java]: nit: remove this empty line?

File: client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java [hunk 14]
```diff
@@ -51,12 +51,18 @@
...
     /**
      * Creates a new builder instance and sets the hosts that the client will send requests to.
+     *
+     * @throws NullPointerException if {@code hosts} or any host is {@code null}.
+     * @throws IllegalArgumentException if {@code hosts} is empty.
      */
     RestClientBuilder(HttpHost... hosts) {
-        if (hosts == null || hosts.length == 0) {
+        Objects.requireNonNull(hosts, "hosts must not be null");
+
```
Comment (javanna) [inline client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java]: remove empty line?

File: client/rest/src/test/java/org/elasticsearch/client/RestClientBuilderTests.java [hunk 15]
```diff
@@ -103,41 +104,80 @@ public void testBuild() throws IOException {
...
             }
             builder.setDefaultHeaders(headers);
         }
-        if (getRandom().nextBoolean()) {
-            builder.setMaxRetryTimeoutMillis(RandomInts.randomIntBetween(getRandom(), 1, Integer.MAX_VALUE));
+        if (randomBoolean()) {
+            builder.setMaxRetryTimeoutMillis(randomIntBetween(1, Integer.MAX_VALUE));
+        }
+        if (randomBoolean()) {
+            String pathPrefix = randomAsciiOfLengthBetween(2, 5);
```
Comment (javanna) [inline client/rest/src/test/java/org/elasticsearch/client/RestClientBuilderTests.java]: can we also randomly prepend pathPrefix with '/' ?

File: client/rest/src/test/java/org/elasticsearch/client/RestClientBuilderTests.java [hunk 16]
```diff
@@ -103,41 +104,80 @@ public void testBuild() throws IOException {
...
+
+            while (pathPrefix.length() < 20 && randomBoolean()) {
+                pathPrefix += "/" + randomAsciiOfLengthBetween(3, 6);
+            }
+
```
Comment (javanna) [inline client/rest/src/test/java/org/elasticsearch/client/RestClientBuilderTests.java]: and randomly append '/' at the end

File: client/rest/src/test/java/org/elasticsearch/client/RestClientBuilderTests.java [hunk 17]
```diff
@@ -103,41 +104,80 @@ public void testBuild() throws IOException {
...
+    private void assertSetPathPrefixThrows(final String pathPrefix) {
+        try {
+            RestClient.builder(new HttpHost("localhost", 9200)).setPathPrefix(pathPrefix);
+            fail("path prefix [" + pathPrefix + "] should have failed");
+        } catch (final IllegalArgumentException e) {
+            assertThat(e.getMessage(), containsString(pathPrefix));
+        }
+    }
+
+
```
Comment (javanna) [inline client/rest/src/test/java/org/elasticsearch/client/RestClientBuilderTests.java]: nit: remove the above two empty lines?

File: client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java [hunk 18]
```diff
@@ -205,6 +206,47 @@ public void testGetWithBody() throws IOException {
...
+        final int statusCode = randomStatusCode(getRandom());
+
+        HttpContext context = null;
+
+        try (final RestClient client =
+                RestClient.builder(new HttpHost(httpServer.getAddress().getHostString(), httpServer.getAddress().getPort()))
+                    .setPathPrefix(pathPrefix).build()) {
+
+            if (rarely() == false) {
+                context = httpServer.createContext("/" + pathPrefix + statusCode + uniqueContextSuffix, new ResponseHandler(statusCode));
```
Comment (javanna) [inline client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java]: I am confused, what happens when rarely is true? don't we need the context in that case?

File: client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java [hunk 19]
```diff
@@ -205,6 +206,47 @@ public void testGetWithBody() throws IOException {
...
+        final String uniqueContextSuffix = "/testPathPrefix";
+        final String pathPrefix = "base/" + randomAsciiOfLengthBetween(1, 5) + "/";
+        final int statusCode = randomStatusCode(getRandom());
+
+        HttpContext context = null;
+
+        try (final RestClient client =
+                RestClient.builder(new HttpHost(httpServer.getAddress().getHostString(), httpServer.getAddress().getPort()))
+                    .setPathPrefix(pathPrefix).build()) {
+
```
Comment (javanna) [inline client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java]: can we test that the request fails before registering the specific context needed for this test?

**Comment by javanna** [comment]
left a couple of minor comments

File: client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java (hunk 18 above)
Comment (pickypg) [inline client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java]: When rarely is true, we don't define a context, so we're testing the behavior without one (HTTP 404s expected).

File: client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java (hunk 19 above)
Comment (pickypg) [inline client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java]: I'll add the context-free part as a separate test. It'll make it a little less convoluted.

**Comment by pickypg** [comment]
@javanna Updated based on comments.

File: client/rest/src/test/java/org/elasticsearch/client/RestClientBuilderTests.java [hunk 20]
```diff
@@ -103,41 +104,77 @@ public void testBuild() throws IOException {
...
+        assertSetPathPrefixThrows("/");
+        assertSetPathPrefixThrows("");
+    }
+
+    public void testSetPathPrefixMalformed() {
+        assertSetPathPrefixThrows("//");
+        assertSetPathPrefixThrows("base/path//");
+    }
+
+    private void assertSetPathPrefixThrows(final String pathPrefix) {
```
Comment (javanna) [inline client/rest/src/test/java/org/elasticsearch/client/RestClientBuilderTests.java]: can we make this static?

File: client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java [hunk 21]
```diff
@@ -205,6 +207,71 @@ public void testGetWithBody() throws IOException {
...
+                try {
+                    esResponse = client.performRequest(method, "/" + statusCode + uniqueContextSuffix);
+                    if ("HEAD".equals(method) == false) {
+                        fail("only HEAD requests should not throw an exception; 404 is expected");
+                    }
+                } catch(ResponseException e) {
+                    esResponse = e.getResponse();
+                }
+
+                assertThat(esResponse.getRequestLine().getUri(), equalTo("/" + pathPrefix + statusCode + uniqueContextSuffix));
```
Comment (javanna) [inline client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java]: you know what? I thin that I was just being paranoid about the test succeeding despite the path prefix not being used. we check that it's used in the returned request line! Maybe we can just remove this method then and keep the following one, which makes much more sense.

Comment (javanna) [inline client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java]: sorry for the confusion!

File: client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java [hunk 22]
```diff
@@ -205,6 +207,71 @@ public void testGetWithBody() throws IOException {
...
+        final String pathPrefix = "base/" + randomAsciiOfLengthBetween(1, 5) + "/";
+        final int statusCode = randomStatusCode(getRandom());
+
+        HttpContext context = null;
+
+        try (final RestClient client =
+                RestClient.builder(new HttpHost(httpServer.getAddress().getHostString(), httpServer.getAddress().getPort()))
+                    .setPathPrefix((randomBoolean() ? "/" : "") + pathPrefix).build()) {
+
+            context = httpServer.createContext("/" + pathPrefix + statusCode + uniqueContextSuffix, new ResponseHandler(statusCode));
```
Comment (javanna) [inline client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java]: can this be initialized earlier, rather than assigning context to null initially?

File: client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java [hunk 23]
```diff
@@ -205,6 +207,71 @@ public void testGetWithBody() throws IOException {
...
+                    esResponse = client.performRequest(method, "/" + statusCode + uniqueContextSuffix);
+                } catch(ResponseException e) {
+                    esResponse = e.getResponse();
+                }
+
+                assertThat(esResponse.getRequestLine().getUri(), equalTo("/" + pathPrefix + statusCode + uniqueContextSuffix));
+                assertThat(esResponse.getStatusLine().getStatusCode(), equalTo(statusCode));
+            }
+        } finally {
+            if (context != null) {
```
Comment (javanna) [inline client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java]: this should always be non null now

**Comment by javanna** [comment]
left a couple minors, LGTM otherwise, feel free to push when those are addressed

File: client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java (hunk 21 above)
Comment (pickypg) [inline client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java]: I'm okay with paranoia in integration testing :). It's good to double check that we're getting the expected response in terms of failure, since otherwise we're only actually testing success. This also doubles as a way to ensure that HEAD 404 is a good response, which I see as a win-win.

File: client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java (hunk 22 above)
Comment (pickypg) [inline client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java]: Good point. I don't actually know why I was doing it that way.

File: client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java (hunk 21 above)
Comment (pickypg) [inline client/rest/src/test/java/org/elasticsearch/client/RestClientIntegTests.java]: Turns out leaving it in was a bad idea. It randomly failed on different JVMs.

c05d5f925789154e62cb0bd94611cd4aedc49755 removed it

**Comment by jfiore-ibm** [comment]
@pickypg Sorry if this is a stupid question, but is there a way to pull down the version of the rest client that has this PR from some Maven repo somewhere?
This PR has the v5.0.0-beta1 label, but I am not sure if that means that there is a 5.0.0-beta1 JAR published to some maven repo, or if it will be whenever that happens.

If the latter, do you have any idea on when that will occur?

Thanks for making this change BTW, I need it :+1:

**Comment by pickypg** [comment]
@jfiore-ibm It should go out in the beta1 release (or any snapshot build) of 5.0.

Fortunately we just released beta1 today. :)

**Comment by jfiore-ibm** [comment]
@pickypg Thanks a bunch! I will pick up the beta1 release.
For my information if I should need it in the future, how do I use a snapshot build (what Maven Repo would I need to use)?
